File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:    
        ancp = []
        ancq = []
        cur = root
        while (cur):
            ancp.append(cur)
            if (cur.val > p.val):
                cur = cur.left
            elif (cur.val < p.val):
                cur = cur.right
            else:
                break
        cur = root
        while (cur):
            ancq.append(cur)
            if (cur.val > q.val):
                cur = cur.left
            elif (cur.val < q.val):
                cur = cur.right
            else:
                break
        i = 0
        lca = None
        while (i < len(ancp) and i < len(ancq) and ancp[i] == ancq[i]):
            lca = ancp[i]
            i += 1
        return lca

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
      int dfsGood(TreeNode *node, int maxSofar) {
    if (!node) {
      return 0;
    }
    int count = 0;
    if (node->val >= maxSofar) {
      count = 1;
      maxSofar = node->val;
    }
    count += dfsGood(node->right, maxSofar);
    count += dfsGood(node->left, maxSofar);

    return count;
  }
  int goodNodes(TreeNode *root) {
    if (!root) {
      return 0;
    }
    return dfsGood(root, root->val);
  }

};
